File: packages/datadog_flutter_plugin/android/src/main/kotlin/com/datadoghq/flutter/DatadogSdkPlugin.kt
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */
package com.datadoghq.flutter

import android.util.Log
import com.datadog.android.Datadog
import com.datadog.android.DatadogSite
import com.datadog.android._InternalProxy
import com.datadog.android.core.configuration.BatchProcessingLevel
import com.datadog.android.core.configuration.Configuration
import com.datadog.android.log.Logs
import com.datadog.android.ndk.NdkCrashReports
import com.datadog.android.privacy.TrackingConsent
import com.datadog.android.rum.GlobalRumMonitor
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.concurrent.ExecutorService
import java.util.concurrent.SynchronousQueue
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit

class DatadogSdkPlugin : FlutterPlugin, MethodCallHandler {
    companion object {
        const val CONTRACT_VIOLATION = "DatadogSdk:ContractViolation"
        const val INVALID_OPERATION = "DatadogSdk:InvalidOperation"
        const val ARG_VALUE = "value"

        // Flutter can destroy / recreate the plugin object if the engine detaches. If you use the
        // back button on the first screen, for example, this will detach the Flutter engine
        // but the application will still be running. We keep the configuration separate
        // from the plugin to warn about reinitialization.
        var previousConfiguration: Map<String, Any?>? = null
    }

    data class ConfigurationTelemetryOverrides(
        var trackViewsManually: Boolean = true,
        var trackInteractions: Boolean = false,
        var trackErrors: Boolean = false,
        var trackNetworkRequests: Boolean = false,
        var trackNativeViews: Boolean = false,
        var trackCrossPlatformLongTasks: Boolean = false,
        var trackFlutterPerformance: Boolean = false,
        var dartVersion: String? = null
    )

    private lateinit var channel: MethodChannel
    private var binding: FlutterPlugin.FlutterPluginBinding? = null

    internal val telemetryOverrides = ConfigurationTelemetryOverrides()

    // Only used to shutdown Datadog in debug builds
    private val executor: ExecutorService = ThreadPoolExecutor(
        0,
        1,
        30L,
        TimeUnit.SECONDS,
        SynchronousQueue()
    )

    internal val logsPlugin: DatadogLogsPlugin = DatadogLogsPlugin()
    internal val rumPlugin: DatadogRumPlugin = DatadogRumPlugin()

    override fun onAttachedToEngine(
        flutterPluginBinding: FlutterPlugin.FlutterPluginBinding
    ) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "datadog_sdk_flutter")
        channel.setMethodCallHandler(this)

        binding = flutterPluginBinding

        logsPlugin.attachToEngine(flutterPluginBinding)
        rumPlugin.attachToEngine(flutterPluginBinding)
    }

    @Suppress("LongMethod")
    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initialize" -> {
                val configArg = call.argument<Map<String, Any?>>("configuration")
                val trackingConsent = call.argument<String>("trackingConsent")?.let {
                    parseTrackingConsent(it)
                }
                if (configArg != null && trackingConsent != null) {
                    if (!Datadog.isInitialized()) {
                        initialize(configArg, trackingConsent)
                        previousConfiguration = configArg
                    } else if (configArg != previousConfiguration) {
                        // Maybe use DevLogger instead?
                        Log.e(DATADOG_FLUTTER_TAG, MESSAGE_INVALID_REINITIALIZATION)
                    }
                    telemetryOverrides.dartVersion = call.argument<String>("dartVersion")
                    result.success(null)
                } else {
                    result.missingParameter(call.method)
                }
            }
            "attachToExisting" -> {
                if (Datadog.isInitialized()) {
                    val attachResult = attachToExisting()
                    result.success(attachResult)
                } else {
                    Log.e(DATADOG_FLUTTER_TAG, MESSAGE_NO_EXISTING_INSTANCE)
                    result.success(null)
                }
            }
            "setSdkVerbosity" -> {
                val value = call.argument<String>(ARG_VALUE)
                if (value != null) {
                    val verbosity = parseCoreLoggerLevel(value)
                    Datadog.setVerbosity(verbosity)
                    result.success(null)
                } else {
                    result.missingParameter(call.method)
                }
            }
            "setTrackingConsent" -> {
                val value = call.argument<String>(ARG_VALUE)
                if (value != null) {
                    val trackingConsent = parseTrackingConsent(value)
                    Datadog.setTrackingConsent(trackingConsent)
                    result.success(null)
                } else {
                    result.missingParameter(call.method)
                }
            }
            "setUserInfo" -> {
                val id = call.argument<String>("id")
                val name = call.argument<String>("name")
                val email = call.argument<String>("email")
                val extraInfo = call.argument<Map<String, Any?>>("extraInfo")
                if (extraInfo != null) {
                    Datadog.setUserInfo(id, name, email, extraInfo)
                    result.success(null)
                } else {
                    result.missingParameter(call.method)
                }
            }
            "addUserExtraInfo" -> {
                val extraInfo = call.argument<Map<String, Any?>>("extraInfo")
                if (extraInfo != null) {
                    Datadog.addUserProperties(extraInfo)
                    result.success(null)
                } else {
                    result.missingParameter(call.method)
                }
            }
            "telemetryDebug" -> {
                val message = call.argument<String>("message")
                if (message != null) {
                    Datadog._internalProxy()._telemetry.debug(message)
                    result.success(null)
                } else {
                    result.missingParameter(call.method)
                }
            }
            "telemetryError" -> {
                val message = call.argument<String>("message")
                if (message != null) {
                    val stack = call.argument<String>("stack")
                    val kind = call.argument<String>("kind")
                    Datadog._internalProxy()._telemetry.error(message, stack, kind)
                    result.success(null)
                } else {
                    result.missingParameter(call.method)
                }
            }
            "updateTelemetryConfiguration" -> {
                updateTelemetryConfiguration(call)
                result.success(null)
            }
            "getInternalVar" -> {
                val name = call.argument<String>("name")
                if (name != null) {
                    val value = getInternalVar(name)
                    result.success(value)
                } else {
                    result.success(null)
                }
            }
            "clearAllData" -> {
                Datadog.clearAllData()
                result.success(null)
            }
            "flushAndDeinitialize" -> {
                invokePrivateShutdown(result)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    fun initialize(encoded: Map<String, Any?>, trackingConsent: TrackingConsent) {
        var configBuilder = configurationBuilderFromEncoded(encoded)
        if (configBuilder == null) {
            return
        }

        binding?.let {
            Datadog.initialize(
                it.applicationContext,
                configBuilder.build(),
                trackingConsent
            )
            (encoded["nativeCrashReportEnabled"] as? Boolean)?.let {
                if (it) {
                    NdkCrashReports.enable()
                }
            }
        }
    }

    @Suppress("ComplexMethod", "FunctionMaxLength")
    fun configurationBuilderFromEncoded(encoded: Map<String, Any?>): Configuration.Builder? {
        val clientToken = encoded["clientToken"] as? String
        val env = encoded["env"] as? String
        if (clientToken == null || env == null) {
            return null
        }

        val service = encoded["service"] as? String
        var needsClearTextHttp = false
        var variant = ""
        (encoded["additionalConfig"] as? Map<String, Any?>)?.let {
            variant = (it["_dd.variant"] as? String) ?: ""
            needsClearTextHttp = (it["_dd.needsClearTextHttp"] as? Boolean) ?: false
        }

        var builder = Configuration.Builder(clientToken, env, variant, service)

        (encoded["site"] as? String)?.let {
            builder = builder.useSite(parseSite(it))
        }

        if (needsClearTextHttp) {
            builder = _InternalProxy.allowClearTextHttp(builder)
        }

        (encoded["batchSize"] as? String)?.let {
            builder = builder.setBatchSize(parseBatchSize(it))
        }
        (encoded["uploadFrequency"] as? String)?.let {
            builder = builder.setUploadFrequency(parseUploadFrequency(it))
        }
        (encoded["batchProcessingLevel"] as? String)?.let {
            builder = builder.setBatchProcessingLevel(parseBatchProcessingLevel(it))
        }
        (encoded["additionalConfig"] as? Map<String, Any>)?.let {
            builder = builder.setAdditionalConfiguration(it)
        }
        (encoded["nativeCrashReportEnabled"] as? Boolean)?.let {
            builder = builder.setCrashReportsEnabled(it)
        }

        return builder
    }

    private fun attachToExisting(): Map<String, Any> {
        val loggingEnabled = Logs.isEnabled()
        val rumEnabled = GlobalRumMonitor.isRegistered()
        if (rumEnabled) {
            rumPlugin.attachToExistingSdk(GlobalRumMonitor.get())
        }

        return mapOf<String, Any>(
            "loggingEnabled" to loggingEnabled,
            "rumEnabled" to rumEnabled
        )
    }

    private fun updateTelemetryConfiguration(call: MethodCall) {
        var isValid = true

        val option = call.argument<String>("option")
        val value = call.argument<Boolean>(ARG_VALUE)
        if (option != null && value != null) {
            when (option) {
                "trackViewsManually" -> telemetryOverrides.trackViewsManually = value
                "trackInteractions" -> telemetryOverrides.trackInteractions = value
                "trackErrors" -> telemetryOverrides.trackErrors = value
                "trackNetworkRequests" -> telemetryOverrides.trackNetworkRequests = value
                "trackNativeViews" -> telemetryOverrides.trackNativeViews = value
                "trackCrossPlatformLongTasks" ->
                    telemetryOverrides.trackCrossPlatformLongTasks = value
                "trackFlutterPerformance" -> telemetryOverrides.trackFlutterPerformance = value
                else -> isValid = false
            }
        } else {
            isValid = false
        }

        if (isValid) {
            rumPlugin.telemetryOverrides = telemetryOverrides
        } else {
            Datadog._internalProxy()._telemetry.debug(
                String.format(MESSAGE_BAD_TELEMETRY_CONFIG, option, value)
            )
        }
    }

    private fun simpleInvokeOn(methodName: String, target: Any) {
        val klass = target.javaClass
        val method = klass.declaredMethods.firstOrNull {
            it.name == methodName || it.name == "$methodName\$dd_sdk_android_core_release"
        }
        method?.let {
            it.isAccessible = true
            it.invoke(target)
        }
    }

    private fun getInternalVar(name: String): Any? {
        return when (name) {
            "mapperPerformance" ->
                mapOf(
                    "total" to mapOf(
                        "minMs" to DatadogRumPlugin.eventMapper.mapperPerf.minInMs,
                        "maxMs" to DatadogRumPlugin.eventMapper.mapperPerf.maxInMs,
                        "avgMs" to DatadogRumPlugin.eventMapper.mapperPerf.avgInMs
                    ),
                    "mainThread" to mapOf(
                        "minMs" to DatadogRumPlugin.eventMapper.mapperPerfMainThread.minInMs,
                        "maxMs" to DatadogRumPlugin.eventMapper.mapperPerfMainThread.maxInMs,
                        "avgMs" to DatadogRumPlugin.eventMapper.mapperPerfMainThread.avgInMs
                    ),
                    "mapperTimeouts" to DatadogRumPlugin.eventMapper.mapperTimeouts
                )
            else -> null
        }
    }

    private fun invokePrivateShutdown(result: Result) {
        executor.submit {
            simpleInvokeOn("flushAndShutdownExecutors", Datadog)
            Datadog.stopInstance()
        }.get()

        result.success(null)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)

        logsPlugin.detachFromEngine()
        rumPlugin.detachFromEngine()
    }
}

internal fun parseTrackingConsent(trackingConsent: String): TrackingConsent {
    return when (trackingConsent) {
        "TrackingConsent.granted" -> TrackingConsent.GRANTED
        "TrackingConsent.notGranted" -> TrackingConsent.NOT_GRANTED
        "TrackingConsent.pending" -> TrackingConsent.PENDING
        else -> TrackingConsent.PENDING
    }
}

internal fun parseSite(site: String): DatadogSite {
    return when (site) {
        "DatadogSite.us1" -> DatadogSite.US1
        "DatadogSite.us3" -> DatadogSite.US3
        "DatadogSite.us5" -> DatadogSite.US5
        "DatadogSite.eu1" -> DatadogSite.EU1
        "DatadogSite.us1Fed" -> DatadogSite.US1_FED
        "DatadogSite.ap1" -> DatadogSite.AP1
        else -> DatadogSite.US1
    }
}

internal fun parseCoreLoggerLevel(level: String): Int {
    return when (level) {
        "CoreLoggerLevel.debug" -> Log.DEBUG
        "CoreLoggerLevel.warn" -> Log.WARN
        "CoreLoggerLevel.error" -> Log.ERROR
        "CoreLoggerLevel.critical" -> Log.ASSERT
        else -> Log.INFO
    }
}

internal fun parseBatchProcessingLevel(level: String): BatchProcessingLevel {
    return when (level) {
        "BatchProcessingLevel.low" -> BatchProcessingLevel.LOW
        "BatchProcessingLevel.medium" -> BatchProcessingLevel.MEDIUM
        "BatchProcessingLevel.high" -> BatchProcessingLevel.HIGH
        else -> BatchProcessingLevel.MEDIUM
    }
}

internal const val DATADOG_FLUTTER_TAG = "DatadogFlutter"

internal const val MESSAGE_INVALID_REINITIALIZATION =
    "🔥 Reinitializing the DatadogSDK with different options, even after a hot restart, is not" +
        " supported. Cold restart your application to change your current configuration."

internal const val MESSAGE_NO_EXISTING_INSTANCE =
    "🔥 attachToExisting was called, but no existing instance of the Datadog SDK exists." +
        " Make sure to initialize the Native Datadog SDK before calling attachToExisting."

internal const val MESSAGE_BAD_TELEMETRY_CONFIG =
    "Attempting to set telemetry configuration option '%s' to '%s', which is invalid."

File: packages/datadog_flutter_plugin/ios/Classes/DatadogSdkPlugin.swift
// Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2019-2022 Datadog, Inc.

import Flutter
import UIKit
import DatadogCore
import DatadogCrashReporting
import DatadogInternal
import DatadogRUM
import DatadogLogs

extension Datadog.Configuration {
    init?(fromEncoded encoded: [String: Any?]) {
        guard let clientToken: String = try? castUnwrap(encoded["clientToken"]),
              let env: String = try? castUnwrap(encoded["env"]) else {
            return nil
        }

        self.init(
            clientToken: clientToken,
            env: env
        )

        service = try? castUnwrap(encoded["service"])

        if let site = convertOptional(encoded["site"], DatadogSite.parseFromFlutter) {
            self.site = site
        }
        if let batchSize = convertOptional(encoded["batchSize"], Datadog.Configuration.BatchSize.parseFromFlutter) {
            self.batchSize = batchSize
        }
        if let uploadFrequency = convertOptional(encoded["uploadFrequency"],
                                                 Datadog.Configuration.UploadFrequency.parseFromFlutter) {
            self.uploadFrequency = uploadFrequency
        }

        _internal_mutation { config in
            config.additionalConfiguration = encoded["additionalConfig"] as? [String: Any] ?? [:]
        }
    }
}

// swiftlint:disable type_body_length
public class DatadogSdkPlugin: NSObject, FlutterPlugin {
    let channel: FlutterMethodChannel

    // NOTE: Although these are instances, they are still registered globally to
    // a method channel. That might be something we want to change in the future
    public private(set) var logs: DatadogLogsPlugin?
    public private(set) var rum: DatadogRumPlugin?

    var currentConfiguration: [AnyHashable: Any]?
    var core: DatadogCoreProtocol?
    var oldConsolePrint: ((String) -> Void)?

    public init(channel: FlutterMethodChannel) {
        self.channel = channel
        super.init()
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "datadog_sdk_flutter", binaryMessenger: registrar.messenger())
        let instance = DatadogSdkPlugin(channel: channel)
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.addApplicationDelegate(instance)
        registrar.publish(instance)

        DatadogLogsPlugin.register(with: registrar)
        DatadogRumPlugin.register(with: registrar)
    }

    // swiftlint:disable:next cyclomatic_complexity function_body_length
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any] else {
            result(
                FlutterError.invalidOperation(message: "No arguments in call to \(call.method)")
            )
            return
        }

        switch call.method {
        case "initialize":
            // swiftlint:disable:next force_cast
            let configArg = arguments["configuration"] as! [String: Any?]
            if let config = Datadog.Configuration.init(fromEncoded: configArg),
               let trackingConsentString: String = try? castUnwrap(arguments["trackingConsent"]) {
                let trackingConsent = TrackingConsent.parseFromFlutter(trackingConsentString)
                if !Datadog.isInitialized() {
                    initialize(configuration: config, trackingConsent: trackingConsent)
                    currentConfiguration = configArg as [AnyHashable: Any]

                    core?.telemetry.configuration(
                        dartVersion: arguments["dartVersion"] as? String
                    )

                    if let crashReportingEnabled = (configArg["nativeCrashReportEnabled"] as? NSNumber)?.boolValue {
                        if crashReportingEnabled {
                            CrashReporting.enable()
                        }
                    }
                } else {
                    let dict = NSDictionary(dictionary: configArg as [AnyHashable: Any])
                    if !dict.isEqual(to: currentConfiguration!) {
                        consolePrint(
                            "🔥 Reinitialziing the DatadogSDK with different options, even after a hot restart," +
                            " is not supported. Cold restart your application to change your current configuation."
                        )
                    }
                }
                result(nil)
            } else {
                result(FlutterError.missingParameter(methodName: call.method))
            }
        case "attachToExisting":
            if Datadog.isInitialized() {
                let attachResult = attachToExisting()
                result(attachResult)
            } else {
                consolePrint(
                    "🔥 attachToExisting was called, but no existing instance of the Datadog SDK exists." +
                    " Make sure to initialize the Native Datadog SDK before calling attachToExisting.")
                result(nil)
            }
        case "setSdkVerbosity":
            if let verbosityString = arguments["value"] as? String {
                let verbosity = CoreLoggerLevel.parseFromFlutter(verbosityString)
                Datadog.verbosityLevel = verbosity
                result(nil)
            } else {
                result(FlutterError.missingParameter(methodName: call.method))
            }
        case "setUserInfo":
            if let extraInfo = arguments["extraInfo"] as? [String: Any?] {
                let id = arguments["id"] as? String
                let name = arguments["name"] as? String
                let email = arguments["email"] as? String
                let encodedAttributes = castFlutterAttributesToSwift(extraInfo)
                Datadog.setUserInfo(id: id, name: name, email: email, extraInfo: encodedAttributes)
                result(nil)
            } else {
                result(FlutterError.missingParameter(methodName: call.method))
            }
        case "addUserExtraInfo":
            if let extraInfo = arguments["extraInfo"] as? [String: Any?] {
                let encodedAttributes = castFlutterAttributesToSwift(extraInfo)
                Datadog.addUserExtraInfo(encodedAttributes)
                result(nil)
            } else {
                result(FlutterError.missingParameter(methodName: call.method))
            }
        case "setTrackingConsent":
            if let trackingConsentString = arguments["value"] as? String {
                let trackingConsent = TrackingConsent.parseFromFlutter(trackingConsentString)
                Datadog.set(trackingConsent: trackingConsent)
                result(nil)
            } else {
                result(FlutterError.missingParameter(methodName: call.method))
            }
        case "telemetryDebug":
            if let message = arguments["message"] as? String {
                Datadog._internal.telemetry.debug(id: "datadog_flutter:\(message)", message: message)
                result(nil)
            } else {
                result(FlutterError.missingParameter(methodName: call.method))
            }
        case "telemetryError":
            if let message = arguments["message"] as? String {
                let stack = arguments["stack"] as? String
                let kind = arguments["kind"] as? String
                Datadog._internal.telemetry.error(id: "datadog_flutter:\(String(describing: kind)):\(message)",
                                                  message: message, kind: kind, stack: stack)
                result(nil)
            } else {
                result(FlutterError.missingParameter(methodName: call.method))
            }
        case "updateTelemetryConfiguration":
            updateTelemetryConfiguration(arguments: arguments)
            result(nil)
        case "getInternalVar":
            guard let varName = arguments["name"] as? String else {
                result(nil)
                return
            }

            let value = getInternalVar(named: varName)
            result(value)
#if DD_SDK_COMPILED_FOR_TESTING
        case "flushAndDeinitialize":
            Datadog.flushAndDeinitialize()
            consolePrint = { value in print(value) }
            logs = nil
            rum = nil
            result(nil)
#endif
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    internal func initialize(configuration: Datadog.Configuration, trackingConsent: TrackingConsent) {
        core = Datadog.initialize(with: configuration, trackingConsent: trackingConsent)
    }

    private func updateTelemetryConfiguration(arguments: [String: Any]) {
        var wasValid = true
        let option = arguments["option"]
        let value = arguments["value"]

        if let option = option as? String,
           let value = value as? Bool {
            switch option {
            case "trackViewsManually":
                core?.telemetry.configuration(
                    trackViewsManually: value
                )
            case "trackInteractions":
                core?.telemetry.configuration(
                    trackInteractions: value
                )
            case "trackErrors":
                core?.telemetry.configuration(
                    trackErrors: value
                )
            case "trackNetworkRequests":
                core?.telemetry.configuration(
                    trackNetworkRequests: value
                )
            case "trackNativeViews":
                core?.telemetry.configuration(
                    trackNativeViews: value
                )
            case "trackCrossPlatformLongTasks":
                core?.telemetry.configuration(
                    trackCrossPlatformLongTasks: value
                )
            case "trackFlutterPerformance":
                core?.telemetry.configuration(
                    trackFlutterPerformance: value
                )
            default:
                wasValid = false
            }
        } else {
            wasValid = false
        }

        if !wasValid {
            Datadog._internal.telemetry.debug(
                id: "datadog_flutter:configuration_error",
                message: "Attempting to set telemetry configuration option '\(String(describing: option))'" +
                    " to '\(String(describing: value))', which is invalid.")
        }
    }

    private func getInternalVar(named name: String) -> Any? {
        switch name {
        case "mapperPerformance":
            let value: [String: Any?] = [
                "total": [
                    "minMs": rum?.mapperPerf.minInMs,
                    "maxMs": rum?.mapperPerf.maxInMs,
                    "avgMs": rum?.mapperPerf.avgInMs
                ],
                "mainThread": [
                    "minMs": rum?.mainThreadMapperPerf.minInMs,
                    "maxMs": rum?.mainThreadMapperPerf.maxInMs,
                    "avgMs": rum?.mainThreadMapperPerf.avgInMs
                ],
                "mapperTimeouts": rum?.mapperTimeouts
            ]
            return value
        default:
            return nil
        }
    }

    public func applicationWillTerminate(_ application: UIApplication) {
        _onDetach()
    }

    public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
        registrar.publish(NSNull())
        _onDetach()
    }

    private func _onDetach() {
        // Reset consolePrint during detach - any other methodChannel callbacks will
        // also need to be cleared as well
        if let oldConsolePrint = oldConsolePrint {
            consolePrint = oldConsolePrint
        }
        oldConsolePrint = nil

        logs?.onDetach()
        rum?.onDetach()
    }

    private func attachToExisting() -> [String: Any?] {
        var rumEnabled = false
        var logsEnabled = false

        core = Datadog.sdkInstance(named: CoreRegistry.defaultInstanceName)
        if Logs._internal.isEnabled() {
            logs = DatadogLogsPlugin.instance
            logsEnabled = true
        }

        if RUM._internal.isEnabled() {
            rum = DatadogRumPlugin.instance
            rum?.attachToExisting(rumInstance: RUMMonitor.shared())
            rumEnabled = true
        }

        return [
            "loggingEnabled": logsEnabled,
            "rumEnabled": rumEnabled
        ]
    }

    // This is a way to work around https://github.com/flutter/flutter/issues/126671
    private func callLogCallback(_ value: String) {
        // Dispatch to the main thread. This is partially to combat Flutter shutdown
        // occuring while we're still processing events
        DispatchQueue.main.async {
            // Second, check that the engine hasn't been destroyed, this is the
            // other work around for https://github.com/flutter/flutter/issues/126671
            if let flutterViewController =
                UIApplication.shared.keyWindow?.rootViewController as? FlutterViewController {
                let engine = flutterViewController.engine
                if engine?.isolateId == nil {
                    return
                }
            }

            self.channel.invokeMethod("logCallback", arguments: value)
        }
    }
}

extension FlutterError {
    public enum DdErrorCodes {
        static let contractViolation = "DatadogSdk:ContractViolation"
        static let invalidOperation = "DatadogSdk:InvalidOperation"
    }

    static func missingParameter(methodName: String) -> FlutterError {
        return FlutterError(code: DdErrorCodes.contractViolation,
                            message: "Missing parameter in call to \(methodName)",
                            details: nil)
    }

    static func invalidOperation(message: String) -> FlutterError {
        return FlutterError(code: DdErrorCodes.invalidOperation,
                            message: message,
                            details: nil)
    }
}
